N°6077 - Attachment class: now creation_date and user_id are init by default

piRGoif opened the pull request:
New fields were added to the class in 2.7.0 with N°330
The creation_date isn't initialized by default, that means every consumer needs to fill it. Currently Mail To Ticket extension doesn't...

This PR adds a `DBInsertNoReload` override in the class so that the field is set by default to the current date.
This would solve part of N°4081

piRGoif committed:
Attachment class : now creation_date is init by default

Hipska commented:
Why not implemented as `$this->SetCurrentDate('creation_date');`?

Or even better, as a lot of other datamodel classes already do, as `$this->SetIfNull('creation_date', time());` in an overload of `OnInsert` or `DBInsertNoReload`.

piRGoif commented:
Indeed you're right, second solution is what is made everywhere in the datamodel ! 
I will switch to it !

piRGoif committed:
Use a DBInsertNoReload overload instead of __construct
Better fit to the usual way this is done (like in the Ticket class for the ref field init)

Thanks @Hipska !

piRGoif committed:
Remove creation_date init in consumer code

Hipska commented on datamodels/2.x/itop-attachments/datamodel.itop-attachments.xml:
```diff
@@ -101,6 +101,18 @@
+    public function DBInsertNoReload()
+    {
+        $this->SetIfNull('creation_date', time());
+        $iKey = parent::DBInsertNoReload();
```
```suggestion
        return parent::DBInsertNoReload();
```

piRGoif replied:
😱
Thanks, I didn't saw that :(

piRGoif committed:
Fix no return for DBInsertNoReload override :(

Thanks again Hipska !

Co-authored-by: Thomas Casteleyn <thomas.casteleyn@super-visions.com>

Molkobain commented on datamodels/2.x/itop-attachments/datamodel.itop-attachments.xml:
```diff
@@ -101,6 +101,18 @@
+          <code><![CDATA[
+    public function DBInsertNoReload()
+    {
+        $this->SetIfNull('creation_date', time());
```
During support/3.0 merge, use the ~`SetComputedDateIfNull()`~ `SetCurrentDateIfNull()` instead.

piRGoif replied:
I didn't find any use of this method and don't understand its phpdoc ? 
What are its benefits ? How should we use it ?

Molkobain replied:
Wrong copy/paste, I meant `SetCurrentDateIfNull()`

Hipska replied:
I also think that might be a better one to use..

piRGoif replied:
Done in 64b7abf3 (target branch is now develop)

Molkobain commented:
Accepted during technical review

Molkobain commented:
Functional review: Pierre will run the audit to list impacted customers. As we tend to not make any datamodel change in a maintenance release, it might be merged in develop only.

piRGoif commented:
Didn't find any Attachment children in the clients customization (details are in N°4081)
Setting this PR to be reviewed for next functional review !

Molkobain commented:
Accepted during functional review but for branch develop instead of support/2.7.
As this is addressing a non blocking bug with a partial fix (other part must be done in the Mail To Ticket Automation extension) we rather stick to our retrofitting guidelines and don't make a datamodel change in a minor version.

To be done before next technical review then merge:
- [x] Add default user_id in the Attachment creation as a fallback in case the consumer didn't pass it.
- [x] Check whereas we need to remove the creation date being set in the portal like it was done for the backoffice
- [x] Rebase PR onto develop branch
- [x] Update PR's title with the bug number in our internal tracker

piRGoif commented:
Rebased branch + PR on develop

piRGoif committed:
Replace SetIfNull with SetCurrentDateIfNull

piRGoif committed:
Also set user if needed

piRGoif committed:
Remove data init in portal
Route attachment/add for \Combodo\iTop\Portal\Controller\ObjectController::AttachmentAction

piRGoif committed:
Remove user_id init for console & portal

piRGoif commented:
PR ready to review again

Molkobain commented:
Accepted during technical review.
You can merge Pierre.

jbostoen commented:
Please also see https://sourceforge.net/p/itop/tickets/2167/
